m3/src/main/java/com/uber/m3/tally/m3/M3Reporter.java

```diff
@@ -556,26 +558,29 @@ public void run() {
                 }
             }
 
-            state.set(ProcessorState.SHUTDOWN);
-
             LOG.warn("Processor shutting down");
 
             shutdown();
+            state.set(ProcessorState.SHUTDOWN);
 
             LOG.warn("Processor shut down");
         }
 
         private void shutdown() {
-            // Drain queue of any remaining metrics submitted prior to shutdown;
-            runNoThrow(this::drainQueue);
-            // Flush remaining buffers at last (best effort)
-            runNoThrow(this::flushBuffered);
+            if (isShutdown.get()) {
+                // If the processor hit an exception, another processor will be created to resume work on the queue,
+                // so leave the queue as is.
+                // Drain queue of any remaining metrics submitted prior to shutdown.
+                runNoThrow(this::drainQueue);
+                // Flush remaining buffers at last (best effort)
+                runNoThrow(this::flushBuffered);
+            }
 
             // Close transport
             transport.close();
 
-            // Count down shutdown latch to notify reporter
-            processorsShutdownLatch.countDown();
+            // Release shutdown semaphore to notify reporter
+            processorsShutdownSema.release();
```

Comment:
Since semaphore acquisition will block the new thread from spinning up, should this release be in a `finally`? Maybe:

```
sema.acquire()
try {
   // whole process method
} finally {
   sema.release()
}
```